src/screencast/screencast_common.c

```diff
@@ -417,3 +401,110 @@ struct xdpw_frame_damage merge_damage(struct xdpw_frame_damage *damage1, struct
 
 	return damage;
 }
+
+void xdpw_buffer_constraints_init(struct xdpw_buffer_constraints *constraints) {
+	*constraints = (struct xdpw_buffer_constraints){ 0 };
+	wl_array_init(&constraints->dmabuf_format_modifier_pairs);
+	wl_array_init(&constraints->shm_formats);
+}
+
+void xdpw_buffer_constraints_finish(struct xdpw_buffer_constraints *constraints) {
+	wl_array_release(&constraints->dmabuf_format_modifier_pairs);
+	wl_array_release(&constraints->shm_formats);
+	if (constraints->gbm) {
+		// TODO: reference count gbm to keep it alive while we have BO's?
+		gbm_device_destroy(constraints->gbm);
+	}
+	*constraints = (struct xdpw_buffer_constraints){ 0 };
+}
+
+static int compare_format_modifier(const void *data_a, const void *data_b) {
+	const struct xdpw_format_modifier_pair *a = data_a;
+	const struct xdpw_format_modifier_pair *b = data_b;
+
+	if (a->fourcc > b->fourcc) {
+		return 1;
+	} else if (a->fourcc < b->fourcc) {
+		return -1;
+	} else if (a->modifier > b->modifier) {
+		return 1;
+	} else if (a->modifier < b->modifier) {
+		return -1;
+	} else {
+		return 0;
+	}
+}
+
+int compare_shm_format(const void *data_a, const void *data_b) {
+	const struct xdpw_shm_format *a = data_a;
+	const struct xdpw_shm_format *b = data_b;
+
+	if (a->fourcc > b->fourcc) {
+		return 1;
+	} else if (a->fourcc < b->fourcc) {
+		return -1;
+	} else {
+		return 0;
+	}
+}
+
+static void xdpw_buffer_constraints_sort(struct xdpw_buffer_constraints *constraints) {
+	qsort(constraints->dmabuf_format_modifier_pairs.data,
+			constraints->dmabuf_format_modifier_pairs.size / sizeof(struct xdpw_format_modifier_pair),
+			sizeof(struct xdpw_format_modifier_pair),
+			compare_format_modifier);
+	qsort(constraints->shm_formats.data,
+			constraints->shm_formats.size / sizeof(struct xdpw_shm_format),
+			sizeof(struct xdpw_shm_format),
+			compare_shm_format);
+}
+
+bool xdpw_buffer_constraints_move(struct xdpw_buffer_constraints *dst, struct xdpw_buffer_constraints *src) {
+	if (!src->dirty) {
+		return false;
+	}
+	int dirty = src->dirty;
+	xdpw_buffer_constraints_sort(src);
```

Comment:
Do we need to sort the formats?

Reply:
Not anymore, no - removed!

(There's technically a gap between ext_image_capture_copy_capture being strictly unordered and pipewire *maybe* being ordered by preference? But let's deal with that if we hit it.)